import fs from 'fs';

import { SignerWithAddress } from '@nomicfoundation/hardhat-ethers/signers';
import { expect } from "chai";
import { ethers } from "hardhat";

const {
  encodeBytes32String,
  decodeBytes32String,
  id,
  getSigners,
  parseUnits,
  dataSlice,
  getBytes,
  concat,
  hexlify,
  keccak256,
  randomBytes,
  Transaction,
  toQuantity,
  toBeHex,
  toUtf8Bytes,
  zeroPadValue } = ethers;

import { CrypticRaffle, SecretDoor } from '../typechain-types';

describe('Sensitive On-Chain Data Exercise 2', () => {
  const provider = ethers.provider;
  let muggle: SignerWithAddress;

  // Original code
  const SECRET_DOOR_ABI = fs.readFileSync('./test/SecretDoorABI.json').toString();

  // Other way to load ABI
  // const rawData = fs.readFileSync('./test/SecretDoor.json', 'utf8');
  // const jsonData = JSON.parse(rawData);
  // const SECRET_DOOR_ABI = jsonData.abi;

  const SECRET_DOOR_ADDRESS = '0x148f340701D3Ff95c7aA0491f5497709861Ca27D';

  let secretDoor: SecretDoor;

  before(async () => {
    [muggle] = await getSigners();

    // Load SecretDoor Contract
    secretDoor = new ethers.Contract(SECRET_DOOR_ADDRESS, SECRET_DOOR_ABI, muggle) as unknown as SecretDoor;

    await secretDoor.unlockDoor(encodeBytes32String('EatSlugs'));
    console.log('encoded bytes: ', encodeBytes32String('EatSlugs'));
  });

  it('Exploit', async () => {
    console.log('encoded bytes: ', encodeBytes32String('Al0h0m0ra'));
    await secretDoor.unlockDoor(encodeBytes32String('Al0h0m0ra'));
  });

  after(async () => {
    console.log("secretDoor.isLocked(): ", await secretDoor.isLocked());
    expect(await secretDoor.isLocked()).equals(false);
  });
});

:   1) Sensitive On-Chain Data Exercise 2
       "after all" hook for "Exploit":
     Error: could not decode result data (value="0x", info={ "method": "isLocked", "signature": "isLocked()" }, code=BAD_DATA, version=6.13.2)
      at makeError (node_modules/ethers/src.ts/utils/errors.ts:694:21)
      at assert (node_modules/ethers/src.ts/utils/errors.ts:715:25)
      at Interface.decodeFunctionResult (node_modules/ethers/src.ts/abi/interface.ts:916:15)
      at staticCallResult (node_modules/ethers/src.ts/contract/contract.ts:346:35)
      at async staticCall (node_modules/ethers/src.ts/contract/contract.ts:303:24)
      at async Proxy.isLocked (node_modules/ethers/src.ts/contract/contract.ts:351:41)
      at async Context.<anonymous> (test/4-sensitive-on-chain-data-2.ts:59:44)
  : why does this kind of error occur although one unit test passed?
  
